### PR TITLE
As a Gitcoin admin, I want an embedded survey for each Gitcoin hackathon so I know the motivations of each hacker

### PR DESCRIPTION
**Pull request by zoek1**
##### Description

- [x] Add models to build dynamic surveys
- [x] Add flow to build a survey from the admin panel

##### Refers/Fixes

#6087 

##### Testing
DEMO: https://www.loom.com/share/9d95aa6549e54e779e9124959c0a0bb7

### Register views
not_registered

## Form to get expectation from the user
Funding The Future! Onboard _ Gitcoin

### The user is registered but the hackathon doesn't start yet
register_no_active

### The user is registered but the hackathon already started
active

## Mobile view 
image

image

image

image

### CODE REVIEW

File: app/dashboard/models.py [hunk 1]
```diff
@@ -4867,3 +4868,64 @@ class TribeMember(SuperModel):
...
+class Answer(SuperModel):
+    user = models.ForeignKey(User, on_delete=models.SET_NULL, null=True, blank=True)
+    question = models.ForeignKey(Question, on_delete=models.SET_NULL, null=True, blank=True)
+    open_response = models.CharField(max_length=350, blank=True, null=True)
+    choice = models.ForeignKey(Option, on_delete=models.CASCADE, null=True, blank=True)
+    created_on = models.DateField(auto_now=True)
+    updated_at = models.DateField(auto_now_add=True)
+
+
+class HackathonOnboarding(SuperModel):
```
Comment (thelostone-mc) [inline app/dashboard/models.py:4924]: Do we need this ? I don't see a need for it yet 
If so -> could we remove this ?

File: app/dashboard/models.py [hunk 2]
```diff
@@ -4867,3 +4868,64 @@ class TribeMember(SuperModel):
...
     )
+
+
+class Poll(SuperModel):
+    POLL_TYPE = (
+        ('HACKATHON', 'Hackathon'),
+    )
+
+    user = models.OneToOneField(User, on_delete=models.SET_NULL, null=True, blank=True)
+    poll_type = models.CharField(max_length=50, choices=POLL_TYPE, blank=False, null=True)
```
[the remark this answers is not in the document]
Comment (zoek1) [reply app/dashboard/models.py:4879]: Sure, is this relevant in the case requires to integrate polls in other views not only in hackathons. This helps to differentiate when is linked to one hackathon or not.

File: app/dashboard/models.py (hunk 1 above)
Comment (zoek1) [reply app/dashboard/models.py]: No, we didn't. I deleted it; :smile:

**Comment by zoek1** [comment]
@thelostone-mc done!

File: app/dashboard/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thelostone-mc) [reply app/dashboard/models.py:4878]: Hmm it seems redundant at this point cause of L4881 where we have a hackathon field so
In a way we establish that this is already have for hackathons

Could we remove this for now and update the structure later if and when the usecase arises ?

**Comment by zoek1** [comment]
Done @thelostone-mc :+1:

File: app/dashboard/templates/dashboard/hackathon/onboard.html [hunk 3]
```diff
@@ -69,64 +96,108 @@
...
-                <button class="btn btn-gc-blue btn-sm d-none" id="save-description-btn" data-savehackathon="{{hackathon.slug}}">save</button>
+          <div class="d-flex justify-content-between col-12">
+              <div class="{% if hackathon.start_date|timesince >= "1 min"%}col-12 {% else %} col-9{% endif %}">
+                {% if hackathon.start_date|timesince >= "1 min"%}
+                <h4 class="font-weight-bolder">
+                  Here are other things you can do
+                </h4>
+                {% else %}
+                <h4 class="font-weight-bolder">
+                  Contratulations! <br> Your registration for  {{ hackathon.name }} is successful.
```
Comment (danlipert) [inline app/dashboard/templates/dashboard/hackathon/onboard.html:137]: typo here

Comment (danlipert) [reply app/dashboard/templates/dashboard/hackathon/onboard.html:137]: spelling and spacing as well

File: app/dashboard/templates/dashboard/hackathon/onboard.html [hunk 4]
```diff
@@ -69,64 +96,108 @@
...
+                </h4>
+                <h4 id="countdown" class="font-weight-bolder pt-4 mt-0">
+                  {{ hackathon.start_date|date:"M j, Y" }} - {{ hackathon.end_date|date:"M j, Y" }}
+                </h4>
+                <p class="pt-4 mb-4">Until then, we recommend a few different ways to prepare:</p>
+                {% endif %}
+                <ul  class="list-unstyled">
+                    <li class="mb-3 gc-text-blue font-weight-bold"><a href="/townsquare?tab=hackathon:{{ hackathon.id }}&trending=0&personal=0"><i class="mr-1 fas fa-check-circle gc-text-blue"></i> Introduce yourself on Town Square + what you like to build (and earn your first Kudos!)</a></li>
+                    <li class="mb-3 gc-text-blue font-weight-bold"><i class="mr-1 fas fa-check-circle gc-text-blue"></i> RSVP our 3 workshops</li>
```
Comment (danlipert) [inline app/dashboard/templates/dashboard/hackathon/onboard.html:146]: should be `RSVP to our 3 workshops`

File: app/dashboard/views.py [hunk 5]
```diff
@@ -4050,6 +4061,27 @@ def hackathon_registration(request):
...
+        if poll:
+            poll = json.loads(poll)
+            set_questions = {}
+            for entry in poll:
+                try:
+                    question = Question.objects.get(id=int(entry['name']))
+                    option = Option.objects.get(id=int(entry['value']))
+
+                    try:
+                        Answer.objects.get(user=request.user, question=question, choice=option)
```
Comment (danlipert) [inline app/dashboard/views.py:4073]: Can use `get_or_create` here instead https://docs.djangoproject.com/en/3.0/ref/models/querysets/#get-or-create

File: app/dashboard/views.py [hunk 6]
```diff
@@ -4050,6 +4061,27 @@ def hackathon_registration(request):
         except Exception as e:
             logger.error('Error while adding to chat', e)
 
+        if poll:
+            poll = json.loads(poll)
+            set_questions = {}
+            for entry in poll:
+                try:
+                    question = Question.objects.get(id=int(entry['name']))
```
Comment (danlipert) [inline app/dashboard/views.py:4069]: I'm guessing that you wrapped this is a try/except to catch if these two querysets fail. Lets use `get_object_or_404` instead

**Comment by PixelantDesign** [comment]
The responsive view is not in this PR, could we roll this out and get responsiv view out next week?

**Comment by zoek1** [comment]
I'll work on the responsive view today, I'll notify here when is done.

**Comment by PixelantDesign** [comment]
Thanks for the updates @zoek1! Responsive views look good to me.

The modal on responsive view has different options than the normal view and design spec....could you double check why that is? See your above screenshots:




Screen Shot 2020-04-28 at 8 21 40 AM

**Comment by zoek1** [comment]
@pixelantdesign it's because the questions aren't hardcoded, any of the admins could add, modify, or remove options from the admin panel. In the video demo, I showed how to create options. The question text and the options are dynamics

**Comment by danlipert** [comment]
Hey @zoek1 - after merging this work I spent some additional time testing it and found a lot of issues, so I'm going to disable the popup and tell our staff not to use this functionality until it is fixed and fully tested. Here are some of the problems I found:

- Adding multiple questions of each type results in layout issues
- Open ended questions do not work
- The modal pops up blank if there is no survey

There may be other problems as well as I feel you did not test this thoroughly. Please submit a followup PR that is well tested and fixes these issues - thanks

**Comment by zoek1** [comment]
@danlipert I mentioned in the video demo that "Open-ended questions" do not work and they don't part of the spec, this is a draft for such type of questions.
Only questions that have multiple options were requested in the bounty, the rest have support only in the backend but not in the frontend that why the issues. 
The modal was my error, I thought based on the figma flow all following hackathons should have a survey so don't display the modal wasn't in my flow.
I tested many times the requested flow, and I mentioned what works and what is a draft. have you checked the video before approving the PR? :thinking:

**Comment by danlipert** [comment]
@zoek1 I see - in the future, don't include any half-completed features if they don't work properly. You have to consider the end user of this functionality who will not be aware what features work and which don't.

**Pull request merged**